### PR TITLE
Add FastForward Before

### PR DESCRIPTION
**Pull request by valentinbonneaud**
Add option to the fastForward function to return the occurrence before the given date

### CODE REVIEW

File: lib/Recur/RRuleIterator.php [hunk 1]
```diff
@@ -167,18 +184,30 @@ public function fastForward(DateTimeInterface $dt)
...
             // We don't know the counter at this point anymore
             $this->counter = NAN;
         }
+
+        if ($before) {
+            $this->currentDate = $previousDate;
+        }
     }
 
+
```
Comment (ThHareau) [inline lib/Recur/RRuleIterator.php]: These lines should be removed

File: tests/VObject/Recur/FastForwardTest.php [hunk 2]
```diff
@@ -69,6 +69,8 @@ public function testFastForwardYearlyBasic()
         $rrule->next();
     }
 
+
```
Comment (ThHareau) [inline tests/VObject/Recur/FastForwardTest.php]: These lines too

File: lib/Recur/RRuleIterator.php [hunk 3]
```diff
@@ -127,37 +127,54 @@ public function isInfinite()
...
-                    case 'yearly':
-                        $i = $diff->days / 365;
-                        break;
-                }
+                $i = $diff->days / $frequencyCoeff;
+
                 $i /= $this->interval;
                 $i /= 4;
                 $i = floor($i);
                 $i = max(1, $i);
```
Comment (ThHareau) [inline lib/Recur/RRuleIterator.php]: `$this#next` is expecting an int. If this library is moved to php 7.1 with strict mode, a cast should be added here

Comment (valentinbonneaud) [reply lib/Recur/RRuleIterator.php]: Corrected, thanks!

File: lib/Recur/RRuleIterator.php [hunk 4]
```diff
@@ -127,55 +127,120 @@ public function isInfinite()
...
+                $frequencyCoeff = 365;
+                break;
+        }
+
+        return $frequencyCoeff;
+    }
+
+    /**
+     * Perform a fast forward by doing jumps based on the distance of the requested date and the frequency of the
+     * recurrence rule. Will return the last occurrence before the requested date.
```
Comment (ThHareau) [inline lib/Recur/RRuleIterator.php]: I don't see a return statement, can you update the doc please?

File: lib/Recur/RRuleIterator.php [hunk 5]
```diff
@@ -127,55 +127,120 @@ public function isInfinite()
...
+
+        isset($previousDate) && $this->currentDate = $previousDate;
+    }
+
+    /**
+     * Return the frequency in number of days.
+     *
+     * @return float|int|null
+     */
+    private function getFrequencyCoeff()
```
Comment (ThHareau) [inline lib/Recur/RRuleIterator.php]: What about moving this switch to an array? All of the value are constants, you could event define a new private const at the top of this file? 

The methods could then be simplified to: 

```
    private function getFrequencyCoeff()
    {
        return static::FREQUENCY_COEFFS[$this->frequency] ?? null;
    }
```

Comment (valentinbonneaud) [reply lib/Recur/RRuleIterator.php]: Const array was introduced in PHP 5.6, we need to support >=5.5. To keep it simple, we can keep the switch

File: lib/Recur/RRuleIterator.php (hunk 4 above)
Comment (valentinbonneaud) [reply lib/Recur/RRuleIterator.php]: Indeed, it's corrected! Thanks

**Pull request merged**